### PR TITLE
refactor(hl7-transformer): replace smolder HL7 data source with native PySpark reader

### PR DESCRIPTION
**Pull request by kathrynalpert**
### Product
N/A — no user-facing change.

### Technical
Replaces the [smolder](https://github.com/washu-tag/smolder) Spark data source (`spark.read.format("hl7")`) with a native PySpark reader (`hl7reader.read_hl7`) — ~30 lines of Catalyst expressions, no UDFs — and drops the Scala 2.12 smolder jar from the transformer image.

Two reasons:
- **Unblocks the Spark 4 upgrade** (next PR in the stack): Spark 4 is Scala 2.13-only, and the smolder jar is `_2.12`.
- **Retires a dormant dependency**: the `washu-tag/smolder` fork's only local patch was already merged upstream, and upstream has no Spark 4 / Scala 2.13 work.

The reader preserves smolder's parsing semantics the pipeline depends on: trailing empty fields dropped / interior kept (Java `String.split` behavior), `message` taken from the first line before empty-line filtering, and `\r` / `\n` / `\r\n` terminators.

## Type of change
- [x] Refactor (code improvement with no functional changes)

##### Authorization
N/A.

##### Appsec
Removes an unmaintained third-party Scala jar from the image (smaller dependency surface).

### Performance
Parity-timed over the staging corpus: native reader within noise of smolder on a full scan; no UDFs, so no Python serialization cost.

### Data
Edge cases verified: trailing vs interior empty fields, blank lines, `\r`/`\n`/`\r\n` terminators, non-ASCII (the extractor writes UTF-8). The only behavioral divergences are on inputs `hl7log-extractor` never produces (zero-byte files, pipes-only segment lines), where smolder previously failed the whole task.

### Backward compatibility
Output parity verified against the smolder jar over all 910 messages from `tests/ingest/staging_test_data` (replicating the extractor's log-split): identical rows, values, and downstream-propagated schema.

## Testing
- New unit tests (`tests/test_hl7reader.py`, 12 cases) covering the parity edge cases above.
- Parity diff old-vs-new over the 910-message corpus: 0 differences.
- CI deploy-and-test exercises the full ingest pipeline end-to-end.

## Note for reviewers
Stacked on #451 (`pr/cve-refresh`) so CI inherits the jackson fix and keycloak ignorefile; the diff against that base is just this one commit. GitHub will auto-retarget the base to `main` once #451 merges. This is the first of the smolder → Spark 4 stack; the Spark 4.1.1 / Delta 4.3.0 upgrade follows in the next PR.

## Checklist
- [x] My code adheres to the coding and style guidelines of the project (ran `pre-commit` on the changed files)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] User documentation — N/A
- [ ] Technical documentation / ADR — N/A
- [x] I have added unit tests
- [ ] End-to-end tests — N/A (existing CI ingest tests cover the path)

### CODE REVIEW

**Comment by kathrynalpert** [comment]
> 🤖 **Review by Claude (Opus 4.8), via Claude Code**

## Verdict: safe — a genuine drop-in replacement

I verified this rather than taking the parity claim on faith: I read smolder's actual Scala source as the baseline (`HL7FileFormat.scala`, `Message.scala`, `Segment.scala`), ran the unit tests in this PR's target config (Java 21 + PySpark 3.5.4 — 12/12 green), and ran my own adversarial probes on edge cases the PR's tests don't cover.

### Is it safe?

Parity is real and tightly bounded:

- **Schema is identical.** `Message.scala` declares exactly `{message: string, segments: array<struct<id: string, fields: array<string>>>}`, and the downstream consumer in `deltalake.py` is unchanged (only a comment changed) — same `filter(segments, x -> x.id = '…')[0].fields`, same `split(message, '\|')`, same `posexplode(obx_lines)`.
- **Field-split semantics match Java/Scala exactly.** Smolder used Scala's `segment.split('|')` (drops trailing empty fields, keeps interior). The reader reproduces this with `regexp_replace(line, '\|+$', '')` then split. I cross-checked the reader against a Java-`String.split` reference over a battery of patterns including `PID|A||B|` and `OBX|1|TX|||x||` — all match.
- **Segment ordering is preserved** (verified) — load-bearing because `report_text` is assembled across repeated `OBX` segments in array order via `posexplode`.
- **Component/repetition separators are left raw** (verified) — the reader correctly does *not* split `^`/`~`; that stays in the unchanged downstream code.
- **Both documented divergences are unreachable in production.** They are cases where smolder *threw and failed the whole Spark task*: zero-byte files and pipes-only segment lines. `SplitHl7LogActivityImpl.java` drops all-blank messages (`allMatch(String::isBlank)`) and writes UTF-8 + `\r`, so the extractor cannot emit a zero-byte `.hl7`; and segment lines always start with an id, so a pipes-only line can't occur.
- **Failure mode actually improves:** where smolder crashed an entire batch on one bad file, the reader tolerates it (or routes it to the existing `write_errors` "not parsable" path via the unchanged `msh-10 is null` check).

### What are we losing?

Effectively nothing in use:

- Smolder's SQL helpers (`parse_hl7_message`, `segment_field`, `subfield`, `repeating_field`) are **not referenced anywhere** in the transformer — only the `format("hl7")` file reader was, and that's what's replaced. Repo-wide sweep finds no dangling references.
- The one real behavioral delta: smolder's contract was "every input file becomes a row, or the task dies loudly"; the reader's is "every *non-empty* file becomes a row; an empty one is silently skipped with no error record." Unreachable today given the extractor guarantee, and even if Spark's `wholetext` empty-file handling ever changed, an empty file would degrade to an error-table row — not data corruption.
- If anything, correctness improves: smolder read with the JVM default charset; the reader pins UTF-8, matching what the extractor writes.

### Will it work on non-WashU HL7?

**Identical to smolder — no regression, no improvement.** The key fact: smolder never read the delimiters from MSH. `Segment.scala` literally has `// assume pipe delimited //@TODO: pull delimiter from MSH segment` and hardcodes `split('|')`. The reader makes the same hardcoded `|` assumption, and component (`^`) / repetition (`~`) / escape handling all live in the unchanged downstream code.

- A non-WashU sender using the HL7 **default** encoding characters (`|`, `^~\&`) — i.e. virtually every real-world HL7v2 feed — works exactly as before.
- A sender using **non-default** field/encoding characters declared in MSH-1/MSH-2 was already unsupported under smolder and remains so. This PR neither introduces nor fixes that.
- Segment terminators `\r` / `\n` / `\r\n` are all handled (a superset matching smolder's `Source.getLines()`), so cross-site terminator variation is fine.

### Minor notes / follow-ups

- **Forward-compat:** the reader uses only stable, ANSI-safe ops (`F.get` returns null out-of-bounds rather than throwing; `slice` length is always ≥ 0), so it should carry cleanly into the stacked Spark 4 PR — which should re-run these tests on 4.1.1 to confirm. I validated on 3.5.4, the version this PR targets.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -144,12 +144,20 @@
       - uses: actions/setup-python@v6.3.0
         with:
           python-version: '3.12'
+      # JVM for the PySpark local-mode tests
+      - uses: actions/setup-java@v5
```
[the remark this answers is not in the document]
Comment (johnflavin) [reply .github/workflows/ci.yaml:148]: I do think we should do this but in a separate PR across all our actions. I'm fine merging this as-is for now.

**Review by johnflavin — APPROVED** [review]
I've deployed this (along with #453) to my dev vm and run an ingest, then verified the outputs as identical to the smolder + spark 3.5.x pipeline. It's ready for CI and a larger validation, which will be easier once merged.

**Comment by cmoore01** [comment]
> I've deployed this (along with #453) to my dev vm and run an ingest, then verified the outputs as identical to the smolder + spark 3.5.x pipeline. It's ready for CI and a larger validation, which will be easier once merged.

Wouldn't we be testing this out on preprod before merging it?

**Comment by johnflavin** [comment]
> > I've deployed this (along with #453) to my dev vm and run an ingest, then verified the outputs as identical to the smolder + spark 3.5.x pipeline. It's ready for CI and a larger validation, which will be easier once merged.
> 
> Wouldn't we be testing this out on preprod before merging it?

I was thinking we would merge first then deploy to preprod. It would be easier to deploy the image as built by CI rather than building locally and pushing, which I had to do for dev. If in those tests we discover something is wrong we can always revert or push another PR with a fix.
